ui-participant/src/providers/I18nProvider.tsx

```diff
@@ -0,0 +1,71 @@
+import React, { createContext, useContext, useEffect, useState } from 'react'
+import { LanguageText } from '@juniper/ui-core'
+import Api from 'api/api'
+import { useUser } from './UserProvider'
+
+const I18nContext = createContext<I18nContextT | null>(null)
+
+export type I18nContextT = {
+  languageTexts: LanguageText[]
+  i18n: (key: string) => string
+}
+
+/**
+ * Returns a hook that allows the caller to internationalize a given piece of text.
+ * Loads the language texts in the selected language.
+ */
+export function useI18n(): I18nContextT {
+  const i18nContext = useContext(I18nContext)
+  if (!i18nContext) {
+    throw ('i18n context not initialized')
+  }
+  return {
+    i18n: i18nContext.i18n,
+    languageTexts: i18nContext.languageTexts
+  }
+}
+
+/**
+ * Provider for the current users i18n context.
+ */
+export default function I18nProvider({ children }: { children: React.ReactNode }) {
+  const { selectedLanguage } = useUser()
+  const [isLoading, setIsLoading] = useState(true)
+  const [isError, setIsError] = useState(false)
+  const [languageTexts, setLanguageTexts] = useState<LanguageText[]>([])
+
+  useEffect(() => {
+    reloadLanguageTexts(selectedLanguage)
+  }, [])
+
+  const reloadLanguageTexts = (selectedLanguage: string) => {
+    setIsLoading(true)
+    Api.getLanguageTexts(selectedLanguage).then(result => {
+      setLanguageTexts(result)
+      setIsError(false)
+      setIsLoading(false)
+    }).catch(() => {
+      setIsError(true)
+      setIsLoading(false)
+    })
+  }
+
+  const i18n = (key: string) => {
+    return languageTexts.find(text => text.messageKey === key)?.text || `{${key}}`
+  }
+
+  return <>
+    {isLoading && <div className="bg-white h-100 w-100">
+      <div className="position-absolute top-50 start-50 translate-middle">Loading...</div>
+    </div>}
```

Comment:
This popup is kinda jarring when switching between languages. Maybe we keep the old language up until the new translation has loaded, so it doesn't the whole screen white. But that might be hard since this is a downstream source of the language.

Reply:
there's something deeply meta about having to have a hardcoded "loading" text on the i18n provider :).    I think you should be able to use the `PageLoadingIndicator` component to do a full screen loading.  Alternatively, I could see just replacing the language selection control text with a loading spinner while the language is loading

Reply:
Agreed that it's a bit hilarious to have the hardcoded text here, of all places.

The issue here is that switching languages doesn't just cause the language texts to reload, it also causes the portal to reload because we need to fetch the site content in the new language. So the main source of the white "Loading..." flash is actually from the PortalProvider. This provider mirrors the portal loading screen which should prevent an _additional_ flash from occurring.

The flash irritated me too but then I remembered that users probably aren't going to be rapidly cycling through languages like I was while testing. I do think we can and should improve loading indicators, but I'd like to keep it outside of the scope of this PR and file a ticket for it, if that's okay.